sdk/core/core-amqp/src/errors.ts

```diff
@@ -0,0 +1,640 @@
+   */
+  constructor(message: string) {
+    super(message);
+  }
+}
+
+/**
+ * Provides a list of retryable AMQP errors.
+ * "InternalServerError", "ServerBusyError", "ServiceUnavailableError", "OperationCancelledError",
+ * "SenderBusyError", "MessagingError", "DetachForcedError", "ConnectionForcedError",
+ * "TransferLimitExceededError"
+ */
+export const retryableErrors: string[] = [
+  "InternalServerError",
+  "ServerBusyError",
+  "ServiceUnavailableError",
+  "OperationCancelledError",
+
+  // OperationTimeoutError occurs when the service fails to respond within a given timeframe.
+  // Since reasons for such failures can be transient, this is treated as a retryable error.
+  "OperationTimeoutError",
+
+  "SenderBusyError",
+  "MessagingError",
+  "DetachForcedError",
+  "ConnectionForcedError",
+  "TransferLimitExceededError",
+
+  // InsufficientCreditError occurs when the number of credits available on Rhea link is insufficient.
+  // Since reasons for such shortage can be transient such as for pending delivery of messages, this is treated as a retryable error.
+  "InsufficientCreditError"
+];
+
+/**
+ * Maps some SytemErrors to amqp error conditions
+ * @enum SystemErrorConditionMapper
+ */
+export enum SystemErrorConditionMapper {
+  ENOTFOUND = "amqp:not-found",
+  EBUSY = "com.microsoft:server-busy",
+  ECONNREFUSED = "amqp:connection:forced",
+  ETIMEDOUT = "com.microsoft:timeout",
+  ECONNRESET = "com.microsoft:timeout",
+  ENETDOWN = "com.microsoft:timeout",
+  EHOSTDOWN = "com.microsoft:timeout",
+  ENETRESET = "com.microsoft:timeout",
+  ENETUNREACH = "com.microsoft:timeout",
+  ENONET = "com.microsoft:timeout"
+}
+
+export function isSystemError(err: any): boolean {
+  let result: boolean = false;
+  if (
+    err.code &&
+    typeof err.code === "string" &&
+    (err.syscall && typeof err.syscall === "string") &&
+    (err.errno && (typeof err.errno === "string" || typeof err.errno === "number"))
+  ) {
+    result = true;
+  }
+  return result;
+}
+
+/**
+ * @internal
+ * Since browser doesnt differentiate between the various kinds of service communication errors,
+ * this utility is used to look at the error target to identify such category of errors.
+ * For more information refer to - https://html.spec.whatwg.org/multipage/comms.html#feedback-from-the-protocol
+ * @param err object that may contain error information
+ */
+function isBrowserWebsocketError(err: any): boolean {
+  let result: boolean = false;
+  if (
+    !isNode &&
+    window &&
+    err.type === "error" &&
+    err.target instanceof (window as any).WebSocket
+  ) {
+    result = true;
+  }
+  return result;
+}
+
+/**
+ * Translates the AQMP error received at the protocol layer or a generic Error into a MessagingError.
+ *
+ * @param {AmqpError} err The amqp error that was received.
+ * @returns {MessagingError} MessagingError object.
+ */
+export function translate(err: AmqpError | Error): MessagingError {
+  if ((err as MessagingError).translated) {
```

Comment:
This part feels a bit confusing. We say that `err` is of type `AmqpError` or `Error` and then immediately cast it to something different than those two options.